File: Kms.Security.Common/Domain/PermissionCategory.cs
using Kms.Security.Util;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Kms.Security.Common.Domain
{
    public class PermissionCategory : IDomainBase<Guid>
    {
        public PermissionCategory()
        {
            PermissionType = Util.PermissionType.Normal;
        }
        public PermissionCategory(Guid id,string name,Guid? parentId,Guid? companyId,PermissionType permissionType)
        {
            Id = id;
            Name = name;
            ParentId = parentId;
            CompanyId = companyId;
            PermissionType = permissionType;
        }
        public void SetParent(PermissionCategory parent)
        {
            Parent = parent;
        }

        public void SetCompany(Company company)
        {
            Company = company;
        }
        public void SetPermissions(IList<Permission> permissions)
        {
            Permissions = permissions;
        }
        public Guid Id { get; protected set; }
        public bool IdIsEmpty()
        {
            return Id == Guid.Empty;
        }
        public void SetNewId()
        {
            Id = Guid.NewGuid();
        }
        public string Name { get; protected set; }

        public Guid? ParentId { get; protected set; }
        public PermissionCategory Parent { get; protected set; }

        public Guid? CompanyId { get; protected set; }
        public Company Company { get; protected set; }
        public PermissionType PermissionType { get; set; }
        public IList<Permission> Permissions { get; protected set; }
    }
}

File: Kms.Security.Portal/app/nodak/common/permissionCategory/permissionCategory.ts
module nodak.common.models {
    export class PermissionCategory extends nodak.models.ModelBase {
        constructor() {
            super(enums.SubSystems.Common, "PermissionCategory");
        }
        AddBussinessModelValidation() { }
        private id: string;
        set Id(value) {
            this.id = value;
        }
        get Id() {
            return this.id;
        }

        private name: string;
        get Name() {
            return this.name;
        }
        set Name(value) {
            this.name = value;
        }

        private parentId: string;
        set ParentId(value) {
            this.parentId = value;
        }
        get ParentId() {
            return this.parentId;
        }

        private parent: PermissionCategory;
        set Parent(value) {
            this.parent = value;
        }
        get Parent() {
            return this.parent;
        }

        private companyId: string;
        get CompanyId() {
            return this.companyId;
        }
        set CompanyId(value) {
            this.companyId = value;
        }

        private company: Company;
        set Company(value) {
            this.company = value;
        }
        get Company() {
            return this.company;
        }

        private permissions: Array<Permission>;
        set Permissions(value) {
            this.permissions = value;
        }
        get Permissions() {
            return this.permissions;
        }
        private permissionType: number;
        get PermissionType() {
            return this.permissionType;
        }
        set PermissionType(value) {
            this.permissionType = value;
        }
    }
}